server/lomas_server/admin_database/mongodb_database.py

```diff
@@ -109,8 +111,13 @@ def get_and_set_may_user_query(self, user_name: str, may_query: bool) -> bool:
         Returns:
             bool: The may_query status of the user before the update.
         """
-        res = self.db.users.find_one_and_update(
-            {"user_name": user_name}, {"$set": {"may_query": may_query}}
+        res = self.db.users.with_options(
+            write_concern=WriteConcern(w=WRITE_CONCERN_LEVEL, j=True)
+        ).find_one_and_update(
+            {"user_name": user_name},
+            {"$set": {"may_query": may_query}},
+            projection={"may_query": 1},
+            return_document=ReturnDocument.BEFORE,
```

Comment:
for my understanding, why do we want it to be returned before?

Reply:
It was already the case by default, I just added the parameter to make it even clearer.
In case of parallel processing, if multiple workers unfortunately call the function at the same time, we want to allow only one to query by user.  Here by reading and updating to False at the same time we avoid the issue (it is an atomic operation: https://medium.com/@AbbasPlusPlus/what-is-atomicity-in-mongodb-ad8f3b4f7d60) if we return a read value, then the may_query is set to False before the program can continue with the returned may_query value.

So here it reads (the value of ReturnDocument.BEFORE), then it uddates may_query (to False), and only then it return the value it has read before updating. With ReturnDocument.AFTER, we would get the value after the update.

Reply:
thanks you!